File: rsa/rsa.cpp
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <ctime>

int gcd(int a, int h) {
    int temp;
    while (1) {
        temp = a % h;
        if (temp == 0)
            return h;
        a = h;
        h = temp;
    }
}

int main() {
    // Input two prime numbers p and q
    int p, q;
    std::cout << "Enter the value of p (a prime number): ";
    std::cin >> p;
    std::cout << "Enter the value of q (a different prime number): ";
    std::cin >> q;

    if (p == q) {
        std::cerr << "p and q must be different prime numbers." << std::endl;
        return 1;
    }

    // First part of public key
    int n = p * q;
    // Finding other part of public key (e for encryption)
    int e = 2;
    int phi = (p - 1) * (q - 1);

    while (e < phi) {
        // e must be co-prime to phi and smaller than phi
        if (gcd(e, phi) == 1)
            break;
        else
            e++;
    }

    // Private key (d for decryption)
    // Choosing d such that it satisfies d * e = 1 + k * phi
    int k = 2; // A constant value
    int d = (1 + (k * phi)) / e;

    // Message to be encrypted
    double msg = 20;

    std::cout << "Message data = " << msg << std::endl;

    // Encryption c = (msg^e) % n
    int c = static_cast<int>(std::pow(msg, e));
    c = c % n;

    std::cout << "Encrypted data = " << c << std::endl;

    // Decryption m = (c^d) % n
    int m = static_cast<int>(std::pow(c, d));
    m = m % n;

    std::cout << "Original Message Sent = " << m << std::endl;

    return 0;
}

File: rsa/rsa.py
import math

# Returns gcd of a and b
def gcd(a, h):
    while True:
        temp = a % h
        if temp == 0:
            return h
        a = h
        h = temp

# Two random prime numbers
p = 3
q = 7

# First part of the public key: n
n = p * q

# Finding the other part of the public key (e stands for encrypt)
e = 2
phi = (p - 1) * (q - 1)

while e < phi:
    # e must be coprime to phi and smaller than phi
    if gcd(e, phi) == 1:
        break
    else:
        e += 1

# Private key (d stands for decrypt)
# choosing d such that it satisfies d * e = 1 + k * totient
k = 2  # A constant value
d = (1 + (k * phi)) // e

# Message to be encrypted
msg = 20
print("Message data =", msg)

# Encryption: c = (msg ^ e) % n
c = pow(msg, e, n)
print("Encrypted data =", c)

# Decryption: m = (c ^ d) % n
m = pow(c, d, n)
print("Original Message Sent =", m)
